bridge/src/connection.rs

```diff
@@ -0,0 +1,150 @@
+use std::convert::TryFrom;
+use temporal_client::{
+    ClientInitError, ClientOptionsBuilder, ClientOptionsBuilderError, WorkflowService, RetryClient,
+    ConfiguredClient, WorkflowServiceClientWithMetrics
+};
+use tonic::Request;
+use url::Url;
+
+pub type Client = RetryClient<ConfiguredClient<WorkflowServiceClientWithMetrics>>;
+
+#[derive(Debug)]
+pub enum ConnectionError {
+    InvalidUrl(url::ParseError),
+    InvalidProtobuf(prost::DecodeError),
+    InvalidConnectionOptions(ClientOptionsBuilderError),
+    InvalidRpc,
+    UnableToConnect(ClientInitError),
+    RequestError(tonic::Status),
+}
+
+impl std::fmt::Display for ConnectionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ConnectionError::InvalidUrl(error) => write!(f, "{}", error),
+            ConnectionError::InvalidProtobuf(error) => write!(f, "{}", error),
+            ConnectionError::InvalidConnectionOptions(error) => write!(f, "{}", error),
+            ConnectionError::InvalidRpc => write!(f, "Invalid RPC call"),
+            ConnectionError::UnableToConnect(error) => write!(f, "{}", error),
+            ConnectionError::RequestError(status) => write!(f, "{}", status)
+        }
+    }
+}
+
+impl From<url::ParseError> for ConnectionError {
+    fn from(err: url::ParseError) -> Self {
+        ConnectionError::InvalidUrl(err)
+    }
+}
+
+impl From<prost::DecodeError> for ConnectionError {
+    fn from(err: prost::DecodeError) -> Self {
+        ConnectionError::InvalidProtobuf(err)
+    }
+}
+
+impl From<ClientOptionsBuilderError> for ConnectionError {
+    fn from(err: ClientOptionsBuilderError) -> Self {
+        ConnectionError::InvalidConnectionOptions(err)
+    }
+}
+
+impl From<ClientInitError> for ConnectionError {
+    fn from(err: ClientInitError) -> Self {
+        ConnectionError::UnableToConnect(err)
+    }
+}
+
+impl From<tonic::Status> for ConnectionError {
+    fn from(err: tonic::Status) -> Self {
+        ConnectionError::RequestError(err)
+    }
+}
+
+impl std::error::Error for ConnectionError {}
+
+
+fn rpc_req<P: prost::Message + Default>(bytes: Vec<u8>) -> Result<tonic::Request<P>, prost::DecodeError> {
+    let proto = P::decode(&*bytes)?;
+    Ok(Request::new(proto))
+}
+
+fn rpc_resp<P: prost::Message + Default>(res: Result<tonic::Response<P>, tonic::Status>) -> Result<Vec<u8>, ConnectionError> {
+    Ok(res?.get_ref().encode_to_vec())
+}
+
+macro_rules! rpc_call {
+    ($client:expr, $rpc:ident, $bytes:expr) => {
+        rpc_resp($client.$rpc(rpc_req($bytes)?).await)
+    };
+}
+
+// A Connection is a Client wrapper for making RPC calls
+#[derive(Debug, Clone)]
+pub struct Connection {
+    client: Client
+}
```

Comment:
This naming feels a little inverted to me. Typically a client owns a connection, not the other way around.

I might just eliminate this and just call it "Client".

Reply:
This was the subject of a lot of discussion during the proposal. While I completely agree (client talks to server, worker may use client), I was overruled with this third concept of a connection.

Reply:
The naming in Core / TS is inconsistent and looks like we're carrying this inconsistency to the Ruby SDK.
The thing is that the raw `Client` in Core is not the same as the `Client` exposed to the language.
In TS we should have probably called `Connection` `Transport`.

For SDKs like Ruby that use only a high level client and a worker I think we can get rid of the whole `Connection` term, in TS I wouldn't and we will probably end up with 3 types of "transports" when we include grpc-web.

The only downside to not exposing a lower level connection is that since `Client` is bound to a namespace it cannot be reused for multiple namespaces (can be worked around) but that's a non issue IMO.